File: backend/src/LastHour.Domain/Entities/Offer.cs
using LastHour.Domain.Common;
using LastHour.Domain.Exceptions;

namespace LastHour.Domain.Entities;

public class Offer : BaseAuditableEntity
{
    public string StoreId { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public decimal OriginalPrice { get; set; }
    public decimal DiscountPrice { get; set; }
    public int OriginalQuantity { get; set; }
    public int RemainingQuantity { get; set; }
    public DateTime ExpiryTime { get; set; }
    public string Category { get; set; } = string.Empty;
    public string? ImageUrl { get; set; }
    public string? ImageUrls { get; set; }
    public bool IsActive { get; set; } = true;
    public bool IsFeatured { get; set; }
    public double Rating { get; set; }
    public int ReviewCount { get; set; }
    public DateTime? PickupStartTime { get; set; }
    public DateTime? PickupEndTime { get; set; }

    public Store Store { get; set; } = null!;
    public ICollection<OrderItem> OrderItems { get; set; } = [];
    public ICollection<Review> Reviews { get; set; } = [];
    public ICollection<Favorite> Favorites { get; set; } = [];
    public ICollection<CartItem> CartItems { get; set; } = [];

    public decimal DiscountPercentage =>
        OriginalPrice > 0
            ? Math.Round((OriginalPrice - DiscountPrice) / OriginalPrice * 100, 1)
            : 0;

    public bool IsExpired => DateTime.UtcNow >= ExpiryTime;
    public bool IsSoldOut => RemainingQuantity <= 0;

    public void DecreaseQuantity(int quantity)
    {
        if (quantity <= 0) throw new ArgumentException("Quantity must be positive");
        if (quantity > RemainingQuantity)
            throw new DomainException($"Only {RemainingQuantity} items remaining for '{Title}'");
        RemainingQuantity -= quantity;
    }

    public void IncreaseQuantity(int quantity)
    {
        if (quantity <= 0) throw new ArgumentException("Quantity must be positive");
        if (RemainingQuantity + quantity > OriginalQuantity)
            throw new DomainException("Cannot exceed original quantity");
        RemainingQuantity += quantity;
    }
}

File: lib/features/offers/domain/entities/offer.dart
class Offer {
  final String id;
  final String title;
  final String description;
  final String storeId;
  final String storeName;
  final String? storeLogoUrl;
  final double originalPrice;
  final double discountPrice;
  final int remainingQuantity;
  final int originalQuantity;
  final DateTime expiryTime;
  final String category;
  final String? imageUrl;
  final List<String> imageUrls;
  final double distance;
  final double rating;
  final int reviewCount;
  final bool isFavorite;
  final bool isActive;
  final DateTime createdAt;

  const Offer({
    required this.id,
    required this.title,
    this.description = '',
    required this.storeId,
    required this.storeName,
    this.storeLogoUrl,
    required this.originalPrice,
    required this.discountPrice,
    required this.remainingQuantity,
    this.originalQuantity = 0,
    required this.expiryTime,
    required this.category,
    this.imageUrl,
    this.imageUrls = const [],
    this.distance = 0,
    this.rating = 0,
    this.reviewCount = 0,
    this.isFavorite = false,
    this.isActive = true,
    required this.createdAt,
  });

  double get discountPercentage => originalPrice > 0
      ? ((originalPrice - discountPrice) / originalPrice * 100)
      : 0;

  bool get isExpired => DateTime.now().isAfter(expiryTime);

  bool get isSoldOut => remainingQuantity <= 0;
}
